R/human_infection.R

```diff
@@ -209,37 +209,57 @@ calculate_infections <- function(
 
     # Convert prob of infected bite to rate
     rate_bitten <- prob_to_rate(prob)
-
-    # Subset humans with potential relapses
-    with_hypnozoites <- variables$hypnozoites$get_index_of(0)$not(TRUE)
-    with_hypnozoites_vector <- with_hypnozoites$to_vector()
+    rate_bitten_complete <- c(rep(0, parameters$human_population))
+    rate_bitten_complete[source_humans$to_vector()] <- rate_bitten
 
     # Get rates/probs of relapse
-    rate_relapse <- variables$hypnozoites$get_values(with_hypnozoites) * parameters$f
+    rate_relapse <- variables$hypnozoites$get_values() * parameters$f
     prob_relapse <- rate_to_prob(rate_relapse)
 
+    # Combine and relative
+    rate_infection_complete <- rate_bitten_complete + rate_relapse
+    relative_rate <- c(rate_bitten_complete/rate_infection_complete)
+
     # Humans with potential new infections
-    potential_new_infections_humans <- source_humans$or(with_hypnozoites)
+    with_hypnozoites <- variables$hypnozoites$get_index_of(0)$not(TRUE)
+    potential_new_infections_humans <- source_humans$copy()$or(with_hypnozoites)
 
     # Sum rates of bite infection and relapses
-    rate_new_infections_vector <- rep(0, parameters$human_population)
-    rate_new_infections_vector[source_vector] <- rate_bitten
-    rate_new_infections_vector[with_hypnozoites_vector] <- rate_new_infections_vector[with_hypnozoites_vector] + rate_relapse
-    rate_new_infections_vector <- rate_new_infections_vector[rate_new_infections_vector>0]
-    # Convert total rate of new infection to probability
-    prob_new_infections_vector <- rate_to_prob(rate_new_infections_vector)
+    prob_new_infections <- rate_to_prob(rate_infection_complete)
 
     # Subset for new infections/bite infections
-    newly_infected <- bitset_at(potential_new_infections_humans, bernoulli_multi_p(prob_new_infections_vector))
+    newly_infected <- bitset_at(potential_new_infections_humans,
+                                bernoulli_multi_p(prob_new_infections[potential_new_infections_humans$to_vector()]))
     newly_bite_infected <- bitset_at(
-      potential_new_infections_humans,
-      bernoulli_multi_p(rate_bitten/rate_new_infections_vector[source_vector]))$and(newly_infected)
+      newly_infected,
+      bernoulli_multi_p(relative_rate[newly_infected$to_vector()]))
```

Comment:
This looks like it's going to slow vivax computations down a lot :(. There are a lot of very slow `$to_vector` operations.

But it's unclear to me how to speed this up on first look, but I'll try again tomorrow.